src/OpenToolkit.Mathematics/Color/Color3.cs

```diff
@@ -0,0 +1,1231 @@
+    /// </summary>
+    /// <typeparam name="T">The color space of the given color.</typeparam>
+    public struct Color3<T> : IEquatable<Color3<T>>
+        where T : IColorSpace3
+    {
+        /// <summary>
+        /// The first element of this color. This corresponds to the first letter of the color type.
+        /// </summary>
+        public float X;
+
+        /// <summary>
+        /// The second element of this colour. This corresponds to the second letter of the colour type.
+        /// </summary>
+        public float Y;
+
+        /// <summary>
+        /// The third element of this color. This corresponds to the second letter of the colour type.
+        /// </summary>
+        public float Z;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Color3{T}"/> struct.
+        /// </summary>
+        /// <param name="x">The first element of the color.</param>
+        /// <param name="y">The second element of the color.</param>
+        /// <param name="z">The third element of the color.</param>
+        public Color3(float x, float y, float z)
+        {
+            X = x;
+            Y = y;
+            Z = z;
+        }
+
+        /// <summary>
+        /// Gets or sets the nth element of this color.
+        /// </summary>
+        /// <param name="n">The nth element to get.</param>
+        public float this[int n]
+        {
+            get
+            {
+                // the old-style switch compiles to slightly smaller IL than the newer one,
+                // so is used here for performance.
+                switch (n)
+                {
+                    case 0:
+                        return X;
+                    case 1:
+                        return Y;
+                    case 2:
+                        return Z;
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(n), n, null);
+                }
+            }
+
+            set
+            {
+                switch (n)
+                {
+                    case 0:
+                        X = value;
+                        break;
+                    case 1:
+                        Y = value;
+                        break;
+                    case 2:
+                        Z = value;
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(n), n, null);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Instantiates a new <see cref="Color3{T}"/> from the X, Y and Z components of a <see cref="Vector3"/>.
+        /// </summary>
+        /// <param name="vec">The vector which the instantiated color will use the X, Y and Z from, in that order.</param>
+        /// <returns>The color that was instantiated.</returns>
+        [Pure]
+        public static explicit operator Color3<T>(Vector3 vec) => new Color3<T>(vec.X, vec.Y, vec.Z);
```

Comment:
Should this vector be marked as `in`? Whatever we do for the speculative `in` PR, we should apply here as well.